src/node/stage/approved.rs

```diff
@@ -1124,7 +1124,13 @@ impl Approved {
             .shared_state
             .compute_relocations(churn_name, churn_signature);
 
-        for (info, action) in relocations {
+        // During startup, members are all at age 4. When a churn_signature has more than 4 trailing
+        // zeros (actually encountered 6 or more quite often), it means all members will be
+        // relocated at the same time, even the elders. So here changed to only relocate the first.
+        // TODO: ideally, it shall be : only one existing/potential elder and multiple adults (not
+        //       include the potential elder if already got one existing elder relocating) can be
+        //       relocated at same time.
+        if let Some((info, action)) = relocations.iter().next() {
```

Comment:
This is both insufficient and unnecessary and also wrong:

**Insufficient**:
Even if we limit the number of simultaneous relocations, what can happen is that some relocations are in progress while another churn event happens which triggers another relocation(s), effectively bypassing any limits.

**Unnecessary**:
- During startup, we don't do relocation, we increment the ages directly. Startup is until the section reaches `recommended_section_size` which is currently 60. So We would need a signature with at least 60 trailing zeroes in order to trigger relocation of all the nodes once the startup phase is over. This is extremely unlikely. 
- In tests the `recommended_section_size` is much lower, but it still shouldn't be a problem in practice. This is because of the "delayed relocation" feature. The way it works is that when we relocate an elder, that elder will keep participating in the section until it is demoted and only then will proceed with the relocation. This will happen even if the elder is actually voted offline. The reason this works is that the relocating elders have incentive to keep participating in the form of a `RelocatePromise` message which they can exchange for actual `Relocate` (which is what increases their age) once they are demoted. 
- Even if we end up relocating all the nodes, the elders will stay in the section until new elders are elected. If there are no nodes to pick the new elders from, the current elders will be blocked from leaving until new nodes join the section and can be promoted to elders

**Wrong**:
We can't limit the relocations to just one at a time, because then we would never be able to progress past the empty prefix. This is because relocations happen only on mature churn (non-infant) and while there is only one section, the only way to cause mature churn is if a mature node leaves the section. So with a limit of only one relocation per churn, that would mean every removed node would cause at most one relocation which means the number of mature nodes would at best stay constant but would actually decrease in practice, therefore never reaching the threshold necessary for split.

Reply:
> So We would need a signature with at least 60 trailing zeroes in order to trigger relocation of all the nodes once the startup phase is over. This is extremely unlikely.

I am not sure about this. As mentioned in the comment, I do observe the situation that a signature with just 6 trailing zeros cause ALL elders got relocated. maybe there is some calculation got mistaken? will double check this

> This is because of the "delayed relocation" feature.

Will the DKG process be affected? The situation here is that: DKG is looking for ALL `after-relocated` nodes as participants

> the elders will stay in the section until new elders are elected

You mean such elder routing node instance keeps two approved instance (i.e. two ids) ?

> if a mature node leaves the section

The current relocation within `prefix ()` happens as `relocate to self with different id`. so that mature node with old id does left the section ?

Reply:
> I am not sure about this. As mentioned in the comment, I do observe the situation that a signature with just 6 trailing zeros cause ALL elders got relocated

I think this is because in tests, `recommended_section_size` is much lower. For example in the `single_split` test it's just 7.

Reply:
> Will the DKG process be affected?

The existing elders, even when already voted offline, will still keep fulfilling their duties as elders. So they will try to select new nodes to promote and will trigger the DKG for them. Then the new nodes will run the DKG and send the results back to the existing elders who will then vote for the new `SectionInfo` and `OurKey` and only when that reaches consensus, will they actually leave, at which point the new nodes will become elders.

Reply:
> You mean such elder routing node instance keeps two approved instance (i.e. two ids) ?

No, they still have their old name. Only after they are demoted will they proceed with the relocation and name change.

Reply:
> I think this is because in tests, recommended_section_size is much lower. For example in the single_split test it's just 7

Then we shall resolve this by changing the testing setup or have some testing specific handling?

> So they will try to select new nodes to promote and will trigger the DKG for them. Then the new nodes will run the DKG. Only after they are demoted will they proceed with the relocation and name change.

But they are actuallly the same physical instance. so at the time that the DKG to be triggered, there is no new nodes to pick up such trigger and starts voting for it. :(